staging/src/k8s.io/apiserver/pkg/util/flowcontrol/apf_controller.go

```diff
@@ -0,0 +1,734 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package flowcontrol
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"math"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	apitypes "k8s.io/apimachinery/pkg/types"
+	apierrors "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/apimachinery/pkg/util/wait"
+	fcboot "k8s.io/apiserver/pkg/apis/flowcontrol/bootstrap"
+	"k8s.io/apiserver/pkg/authentication/user"
+	"k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/apiserver/pkg/util/apihelpers"
+	fq "k8s.io/apiserver/pkg/util/flowcontrol/fairqueuing"
+	fcfmt "k8s.io/apiserver/pkg/util/flowcontrol/format"
+	"k8s.io/apiserver/pkg/util/flowcontrol/metrics"
+	kubeinformers "k8s.io/client-go/informers"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+
+	fctypesv1a1 "k8s.io/api/flowcontrol/v1alpha1"
+	fcclientv1a1 "k8s.io/client-go/kubernetes/typed/flowcontrol/v1alpha1"
+	fclistersv1a1 "k8s.io/client-go/listers/flowcontrol/v1alpha1"
+)
+
+// This file contains a simple local (to the apiserver) controller
+// that digests API Priority and Fairness config objects (FlowSchema
+// and PriorityLevelConfiguration) into the data structure that the
+// filter uses.  At this first level of development this controller
+// takes the simplest possible approach: whenever notified of any
+// change to any config object, or when any priority level that is
+// undesired becomes completely unused, all the config objects are
+// read and processed as a whole.
+
+// To support synchronization in testing, the controller keeps track
+// of the latest priority level referenced by a FlowSchema of this
```

Comment:
outdated comment; what's tracked is the MatchingPriority rather than the PL reference.